rslib/src/cloze.rs

```diff
@@ -31,99 +19,262 @@ lazy_static! {
+                    _ => {}
+                }
             }
-        } else {
-            text_attr = "".to_string();
-            replacement = text;
         }
+    }
 
-        format!(r#"<span class="cloze"{}>{}</span>"#, text_attr, replacement)
-    });
+    if !current_found {
+        return Cow::Borrowed("");
+    }
+    if cloze_only {
+        return stack.last().unwrap().text[2..].to_owned().into();
+    }
 
-    if !cloze_ord_was_in_text {
-        return "".into();
+    if stack.len() > 1 {
+        let cloze = stack.pop().unwrap();
+        stack.last_mut().unwrap().text.push_str(&cloze.buf);
+    }
+    return Cow::from(stack.last().unwrap().text.to_owned());
+
+    // 3 consecutive {{{, shift the first onto the "parent" text
+    fn shift_open(stack: &mut Vec<Cloze>) {
+        let i = stack.len() - 1;
+        let fc = stack[i].buf.remove(0);
+        stack[i - 1].buf.push(fc);
+    }
+
+    // Close cloze and set state
+    fn close(
+        state: &mut State,
+        stack: &mut Vec<Cloze>,
+        cloze_only: bool,
+        cloze_ord: u16,
+        question: bool,
+    ) -> bool {
+        let cloze = stack.pop().unwrap();
+        *state = if stack.len() > 1 {
+            State::Text
+        } else {
+            State::Root
+        };
+
+        let ordinal = cloze.ord_str.parse::<u16>().unwrap();
+        let last = stack.last_mut().unwrap();
+
+        match (
+            cloze_only,
+            question,
+            ordinal == cloze_ord,
+            cloze.hint.is_empty(),
+        ) {
+            // Cloze text only
+            (true, true, true, true) => last.text.push_str(", ..."),
+            (true, true, true, false) => last.text.push_str(&format!(", {}", &cloze.hint)),
+            (true, _, _, _) => last.text.push_str(&format!(", {}", &cloze.text)),
+
+            // Full cloze
+            // Question - active cloze, no hint
+            (false, true, true, true) => last.text.push_str(&format!(
+                r#"<span class="cloze active" data-cloze="{}" data-ordinal="{}">[...]</span>"#,
+                encode_attribute(cloze.text.as_str()),
+                ordinal
+            )),
+            // Question - active cloze, hint
+            (false, true, true, false) => last.text.push_str(&format!(
+                r#"<span class="cloze active" data-cloze="{}" data-ordinal="{}">[{}]</span>"#,
+                encode_attribute(cloze.text.as_str()),
+                ordinal,
+                &cloze.hint
+            )),
+            // Question - inactive cloze
+            (false, true, false, _) => last.text.push_str(&format!(
+                r#"<span class="cloze" data-ordinal="{}">{}</span>"#,
+                ordinal,
+                cloze.text.as_str()
+            )),
+            // Answer - active cloze
+            (false, false, true, _) => last.text.push_str(&format!(
+                r#"<span class="cloze active" data-ordinal="{}">{}</span>"#,
```

Comment:
While a separate active class on only the active clozes may make sense in a new implementation, I'm not sure we can do that here - most users will have lines like `.cloze { color: blue; }` in their card templates, and this change will lead to the inactive clozes to be colored as well, which I'm not sure we want. Perhaps an approach like #2140 might better preserve the existing behaviour?

Reply:
Ok, so how about `class="cloze"` for active and `class="cloze-inactive"` for inactive?

Reply:
Yep, that's less likely to break things